test/integration_test/node_pdb_test.go

```diff
@@ -0,0 +1,205 @@
+		},
+		TestFunc: NodePDBDisablingParallelUpgrade,
+	},
+}
+
+func CreateNodePDBBasic(tc *types.TestCase) func(*testing.T) {
+	return func(t *testing.T) {
+		testSpec := tc.TestSpec(t)
+		cluster, ok := testSpec.(*corev1.StorageCluster)
+		require.True(t, ok)
+		cluster = ci_utils.DeployAndValidateStorageCluster(cluster, ci_utils.PxSpecImages, t)
+		pxVersion := testutil.GetPortworxVersion(cluster)
+
+		if pxVersion.GreaterThanOrEqual(pxVer3_1_2) {
+			logrus.Infof("Validating Node PDB names and default minAvailable")
+			err := testutil.ValidateNodePDB(cluster, ci_utils.DefaultValidateUpgradeTimeout, ci_utils.DefaultValidateUpgradeRetryInterval)
+			require.NoError(t, err)
+		}
+		ci_utils.UninstallAndValidateStorageCluster(cluster, t)
+	}
+
+}
+func CreateNodePDBWithStoragelessNode(tc *types.TestCase) func(*testing.T) {
+	return func(t *testing.T) {
+		testSpec := tc.TestSpec(t)
+		cluster, ok := testSpec.(*corev1.StorageCluster)
+		require.True(t, ok)
+
+		*cluster.Spec.CloudStorage.MaxStorageNodesPerZone = uint32(3)
+		logrus.Info("Validating PDB with storageless nodes using maxstoragenodesperzone value: ", *cluster.Spec.CloudStorage.MaxStorageNodesPerZone)
+		cluster = ci_utils.DeployAndValidateStorageCluster(cluster, ci_utils.PxSpecImages, t)
+		ci_utils.UninstallAndValidateStorageCluster(cluster, t)
+
+	}
+}
+
+func MaxNodesAvailableForUpgrade(tc *types.TestCase) func(*testing.T) {
+	return func(t *testing.T) {
+		testSpec := tc.TestSpec(t)
+		cluster, ok := testSpec.(*corev1.StorageCluster)
+		require.True(t, ok)
+		cluster = ci_utils.DeployAndValidateStorageCluster(cluster, ci_utils.PxSpecImages, t)
+		pxVersion := testutil.GetPortworxVersion(cluster)
+
+		if pxVersion.GreaterThanOrEqual(pxVer3_1_2) {
+			err := ci_utils.CordonNodes()
+			require.NoError(t, err)
+
+			logrus.Infof("Validating number of nodes ready for upgrade without minAvailable annotation")
+			err = testutil.ValidateNodesSelectedForUpgrade(cluster, -1, ci_utils.DefaultValidateUpgradeTimeout, ci_utils.DefaultValidateUpgradeRetryInterval)
+			require.NoError(t, err)
+
+			k8snodecount, err := ci_utils.GetNonMasterK8sNodeCount()
+			require.NoError(t, err)
+			cluster, err = operator.Instance().GetStorageCluster(cluster.Name, cluster.Namespace)
+			require.NoError(t, err)
+			cluster.Annotations["portworx.io/storage-pdb-min-available"] = fmt.Sprintf("%d", k8snodecount-1)
+			cluster, err = ci_utils.UpdateStorageCluster(cluster)
+			require.NoError(t, err)
+
+			logrus.Infof("Validating number of nodes ready for upgrade with minAvailable annotation %d", k8snodecount-1)
+			err = testutil.ValidateNodesSelectedForUpgrade(cluster, k8snodecount-1, ci_utils.DefaultValidateUpgradeTimeout, ci_utils.DefaultValidateUpgradeRetryInterval)
+			require.NoError(t, err)
+
+			err = ci_utils.UncordonNodes()
+			require.NoError(t, err)
+		}
+		ci_utils.UninstallAndValidateStorageCluster(cluster, t)
+	}
+}
+
+func NodePDBDisablingParallelUpgrade(tc *types.TestCase) func(*testing.T) {
```

Comment:
What are we trying to test here?  Are we just checking that the number is less that k8snodecount - 1? This would mean that we will wait for the entire upgrade timeout to validate that. That would be a long running test.

Reply:
All the tests in the operator do not actually perform the k8s upgrade. This just checks the nodes that are selected for upgrade. Since the operator can only control this and the nodes actually selected are not under its control.